Issue #9993: Created util method in JavadocUtil.java to get content of custom tag

Vyom-Yadav opened the pull request:
Resolves #9993: Created util method in JavadocUtil.java to get content of custom tag

Vyom-Yadav commented:
@romani I am having little trouble in creating test cases for the method.

- I can do it by adding every element of the javadoc AST manually by setting the node to a particular type, something like 
```
final JavadocNodeImpl node = new JavadocNodeImpl();
node.setType(theType)
```
for all the elements in the AST, but I don't think this is the best way to do it, isn't there any way in which I can set all the nodes?

AkMo3 commented:
@Vyom-Yadav I a better way according to me is to create a new file, parse the file along with Javadoc comments, and use the returned AST of the first node of the file. Also, you can check your implementation if it is able to replace the current implementation of the following method with your change.

https://github.com/checkstyle/checkstyle/blob/df5cda9b438f8430420d51807c18d1ce51ac489a/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java#L444-L454

to something like 
```
private String getInlineSummary(DetailAST javadoc) {
            return getContentOfSummaryTag(getInlineTag(javadoc));  
}
```

AkMo3 commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -417,4 +417,40 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+        ArrayList<String> summaryTagContent = null;
+        for (DetailNode node : javadoc.getChildren()) {
+            if (node.getType() == JavadocTokenTypes.CUSTOM_NAME
+                    && node.getText().equals("@summary")
```
Please remove the check for `@summary` text, we want this util method to be universal for all `CUSTOM_NAME` tags.

Vyom-Yadav replied:
Checkout my comment on the issue once, I want to make sure I have implemented what is required.

Vyom-Yadav replied:
done

AkMo3 replied:
Resolved.

AkMo3 replied:
Please see updated https://github.com/checkstyle/checkstyle/issues/9993#issuecomment-845768722 and resolve this review from your side.

Vyom-Yadav commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtilTest.java:
```diff
@@ -379,6 +391,36 @@ public void testEscapeAllControlChars() {
+        final DetailNode node = detailNodeParser
+                .parseJavadocAsDetailNode(javadocCommentAST).getTree();
+        final String actual = JavadocUtil.getContentOfInlineCustomTag(node.getChildren()[5]);
+        final String expected = " This is the summary of the class.\n"
```
I added whitespace before "This" but I feel it's not needed, in that case just comment not needed and I will remove it.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,47 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+        else {
+            for (DetailNode child : children) {
+                if (child.getType() != JavadocTokenTypes.LEADING_ASTERISK) {
+                    result = result.concat(extractContentInlineCustomTag(child));
```
please use StringBuilder for such cases, string concatenation in loop is not a good idea
https://stackoverflow.com/questions/1532461/stringbuilder-vs-string-concatenation-in-tostring-in-java

Vyom-Yadav replied:
Noted, done.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtilTest.java:
```diff
@@ -379,6 +391,36 @@ public void testEscapeAllControlChars() {
+                + "InputJavadocUtilTest_ContentOfInlineCustomTag.java");
+        final DetailAST root = JavaParser.parseFile(testFile, JavaParser.Options.WITH_COMMENTS);
+        final DetailAST javadocCommentAST = root.getLastChild().getFirstChild().getFirstChild()
+                .getNextSibling();
```
I do not understand what is going on here. We parse file and then adding something to AST? Why not put everything in file?

Vyom-Yadav replied:
>We parse file and then adding something to AST? Why not put everything in file?

We are not adding something to the ast, the util method takes DetailNode of type `JavadocTokenTypes.JAVADOC_INLINE_TAG` as input, so we parse the file, get the root node, then from the root node we traverse to the required node and perform checking.

>If we need only ast node, why do we need a file in this case?

We need the file otherwise I would have to set text and everything else explicitly which won't be good imo.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,47 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+
+        while (nodeToBeProcessed != endingNode) {
+            customTagContent = customTagContent
+                    .concat(extractContentInlineCustomTag(nodeToBeProcessed));
```
same thing with concatenation here

Vyom-Yadav replied:
done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,46 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+        DetailNode nodeToBeProcessed = childrenOfJavadocInlineTag[2];
+        final DetailNode endingNode =
+                childrenOfJavadocInlineTag[childrenOfJavadocInlineTag.length - 1];
+        final StringBuilder customTagContent = new StringBuilder(400);
```
you can use just one StringBuilder. Create it here and pass it to extractContentInlineCustomTag method instead of creating new StringBuilder every time method is invoked.

Vyom-Yadav replied:
done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,46 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+    private static String extractContentInlineCustomTag(DetailNode node) {
+        final StringBuilder result = new StringBuilder(160);
+        final DetailNode[] children = node.getChildren();
+        if (children.length == 0) {
```
`node.hasChildren()`

Vyom-Yadav replied:
It's not `DetailAst` but can create such a method for `DetailNode` if you want.

strkkk replied:
sorry, I didnt see it is different type

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,44 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+        DetailNode nodeToBeProcessed = childrenOfJavadocInlineTag[2];
+        final DetailNode endingNode =
+                childrenOfJavadocInlineTag[childrenOfJavadocInlineTag.length - 1];
+        final StringBuilder customTagContent = new StringBuilder(520);
```
one minor:
520 seems like a magic number and I do not get why it is like this. 
Such number should be either justified, or just use parameterless ctor with default value.

Vyom-Yadav replied:
I used parameterless initially, TeamCity complained about it so had to give it a number.
> 520 seems like a magic number

I feel the same but I had some input https://stackoverflow.com/questions/13360229/most-efficient-initial-capacity-size-for-stringbuilder (didn't test it at my end)

strkkk replied:
I see that in other places in code mostly 256 or 1024 used. Lets make it 256

Vyom-Yadav replied:
done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,44 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+    private static void extractContentInlineCustomTag(DetailNode node,
+                                                      StringBuilder customTagContent) {
+        final DetailNode[] children = node.getChildren();
+        if (children.length == 0) {
```
so, it is either node text, or its children texts. Is there some case where there is a text for node and it has children?

Vyom-Yadav replied:
Not that I could find one, mostly all nodes that have children do not contain any text.
@nmancus1 Is there any node that follows this kind of pattern- (Some text other than node's name)

```
DESCRIPTION -> Some text other than node's name
 |--TEXT -> link to  
 |--JAVADOC_INLINE_TAG -> Some text other than node's name
 |   |--JAVADOC_INLINE_TAG_START -> { 
 |   |--LINK_LITERAL -> @link 
 |   |--WS -> 
 |   |--REFERENCE -> Some text other than node's name
 |   |   |--PACKAGE_CLASS -> java.lang.Terminator 
 |   |   |--HASH -> # 
 |   |   |--MEMBER -> equals 
 |   |   `--PARAMETERS -> Some text other than node's name
 |   |       |--LEFT_BRACE -> ( 
 |   |       |--ARGUMENT -> java.lang.Object 
 |   |       `--RIGHT_BRACE -> ) 
 |   `--JAVADOC_INLINE_TAG_END -> } 
 `--TEXT ->    
```

nmancus1 replied:
> Is there any node that follows this kind of pattern...

Relevant grammar from `JavadocParser.g4`:

```antlr

description: (
                  ({!isNextJavadocTag()}? LEADING_ASTERISK)
                  | htmlComment
                  | NEWLINE
                  | text
                  | javadocInlineTag
                  | htmlElement
            )+;

```

It appears that this is possible; whether or not it can happen in practice is another story. The best way to find out would be to write such a Javadoc and parse it.

Vyom-Yadav replied:
@strkkk @nmancus1 I don't think it will be possible as everything parses as text inside of inline tag eg-
```
/**
 * {@sumaary
 * <!-- This is a HTML comment -->
 *  <ul>
 *   <li>Coffee</li>
 * </ul>
 * **** some stars
 * }
 * <!-- HTML comment outside the tag -->
 *  <ul>
 *   <li>Coffee</li>
 * </ul>
 */
```
Will parse as-
```
JAVADOC -> JAVADOC [0:0]
|--TEXT -> /** [0:0]
|--NEWLINE -> \n [0:3]
|--LEADING_ASTERISK ->  * [1:0]
|--TEXT ->   [1:2]
|--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [1:3]
|   |--JAVADOC_INLINE_TAG_START -> { [1:3]
|   |--CUSTOM_NAME -> @sumaary [1:4]
|   |--NEWLINE -> \n [1:12]
|   |--LEADING_ASTERISK ->  * [2:0]
|   |--WS ->   [2:2]
|   |--DESCRIPTION -> DESCRIPTION [2:3]
|   |   |--TEXT -> <!-- This is a HTML comment --> [2:3]
|   |   |--NEWLINE -> \n [2:34]
|   |   |--LEADING_ASTERISK ->  * [3:0]
|   |   |--TEXT ->   <ul> [3:2]
|   |   |--NEWLINE -> \n [3:8]
|   |   |--LEADING_ASTERISK ->  * [4:0]
|   |   |--TEXT ->    <li>Coffee</li> [4:2]
|   |   |--NEWLINE -> \n [4:20]
|   |   |--LEADING_ASTERISK ->  * [5:0]
|   |   |--TEXT ->  </ul> [5:2]
|   |   |--NEWLINE -> \n [5:8]
|   |   |--LEADING_ASTERISK ->  * [6:0]
|   |   |--TEXT ->  **** some stars [6:2]
|   |   |--NEWLINE -> \n [6:18]
|   |   |--LEADING_ASTERISK ->  * [7:0]
|   |   `--TEXT ->   [7:2]
|   `--JAVADOC_INLINE_TAG_END -> } [7:3]
|--NEWLINE -> \n [7:4]
|--LEADING_ASTERISK ->  * [8:0]
|--TEXT ->   [8:2]
|--HTML_COMMENT -> HTML_COMMENT [8:3]
|   |--HTML_COMMENT_START -> <!-- [8:3]
|   |--TEXT ->  HTML comment outside the tag  [8:7]
|   `--HTML_COMMENT_END -> --> [8:37]
|--NEWLINE -> \n [8:40]
|--LEADING_ASTERISK ->  * [9:0]
|--TEXT ->    [9:2]
|--HTML_ELEMENT -> HTML_ELEMENT [9:4]
|   `--HTML_TAG -> HTML_TAG [9:4]
|       |--HTML_ELEMENT_START -> HTML_ELEMENT_START [9:4]
|       |   |--START -> < [9:4]
|       |   |--HTML_TAG_NAME -> ul [9:5]
|       |   `--END -> > [9:7]
|       |--NEWLINE -> \n [9:8]
|       |--LEADING_ASTERISK ->  * [10:0]
|       |--TEXT ->     [10:2]
|       |--HTML_ELEMENT -> HTML_ELEMENT [10:5]
|       |   `--LI -> LI [10:5]
|       |       |--LI_TAG_START -> LI_TAG_START [10:5]
|       |       |   |--START -> < [10:5]
|       |       |   |--LI_HTML_TAG_NAME -> li [10:6]
|       |       |   `--END -> > [10:8]
|       |       |--TEXT -> Coffee [10:9]
|       |       `--LI_TAG_END -> LI_TAG_END [10:15]
|       |           |--START -> < [10:15]
|       |           |--SLASH -> / [10:16]
|       |           |--LI_HTML_TAG_NAME -> li [10:17]
|       |           `--END -> > [10:19]
|       |--NEWLINE -> \n [10:20]
|       |--LEADING_ASTERISK ->  * [11:0]
|       |--TEXT ->   [11:2]
|       `--HTML_ELEMENT_END -> HTML_ELEMENT_END [11:3]
|           |--START -> < [11:3]
|           |--SLASH -> / [11:4]
|           |--HTML_TAG_NAME -> ul [11:5]
|           `--END -> > [11:7]
|--NEWLINE -> \n [11:8]
|--LEADING_ASTERISK ->  * [12:0]
|--TEXT -> / [12:2]
`--EOF -> <EOF> [12:3]
```
They are parsed as HTML tags outside the inline tag, inside they are just parsed as text, also outside the inline tag the parent nodes do not have any text which is not contained by the nested/children nodes.

strkkk replied:
ok, let it be like this, in case of some case we didnt expect, fix will be easy

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/utils/javadocutil/InputJavadocUtilTest_ContentOfInlineCustomTag.java:
```diff
@@ -0,0 +1,21 @@
+ * <li>This class is a test class.</li>
+ * </ul>}
+ */
+public class InputJavadocUtilTest_ContentOfInlineCustomTag { // ok
```
usually, naming is like `InputJavadocUtilInlineCustomTag` or something like this. No underscore, and short "description"

Please also add a bit more inputs or extend this one.

Vyom-Yadav replied:
> usually, naming is like Inp......

This PR's quite old, don't remember the rationale for giving it such a complicated name, changed now and added some more inputs

nmancus1 commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtilTest.java:
```diff
@@ -379,6 +389,64 @@ public void testEscapeAllControlChars() {
 
+    @Test
+    public void testGetContentOfInlineCustomTag() throws Exception {
+        final File testFile = new File("src/test/resources/com/"
+                + "puppycrawl/tools/checkstyle/utils/javadocutil/"
+                + "InputJavadocUtilInlineCustomTag.java");
```
Why not use `AbstractPathTestSupport#getPath` here?

Vyom-Yadav replied:
Done

nmancus1 replied:
Resolved.

nmancus1 commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtilTest.java:
```diff
@@ -379,6 +389,64 @@ public void testEscapeAllControlChars() {
     }
 
+    @Test
+    public void testGetContentOfInlineCustomTag() throws Exception {
```
Can we just use multiple input files and break this into multiple test methods or use a parameterized test? To me, in most cases, a test should fail for one reason only. This would also help with complicated traversal of AST.

Vyom-Yadav replied:
Spilt it into two tests, please take a look

nmancus1 replied:
Why can't we have four, each with one expected, one actual, and one assertion?

Vyom-Yadav replied:
Done

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,44 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+     * @param javadocInlineTag javadoc inline tag node.
+     * @return String consisting of the content of inline custom tag.
+     */
+    public static String getContentOfInlineCustomTag(DetailNode javadocInlineTag) {
```
can't we just call this `inlineTag` instead of `javadocInlineTag`?

Vyom-Yadav replied:
Done

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,44 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+     * @return String consisting of the content of inline custom tag.
+     */
+    public static String getContentOfInlineCustomTag(DetailNode javadocInlineTag) {
+        final DetailNode[] childrenOfJavadocInlineTag = javadocInlineTag.getChildren();
```
`childrenOfInlineTag`?

Vyom-Yadav replied:
Done

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,44 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+     * @param node DetailNode
+     * @param customTagContent content of custom tag
+     */
+    private static void extractContentInlineCustomTag(DetailNode node,
```
Since this is a private method used as helper for `getContentOfInlineCustomTag`, can we just call it `extractContent`? Or maybe `extractInlineTagContent`?

Vyom-Yadav replied:
Renamed to `extractInlineTagContent`

nmancus1 replied:
Resolved.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java:
```diff
@@ -408,4 +408,44 @@ else if (sibling.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+     */
+    public static String getContentOfInlineCustomTag(DetailNode javadocInlineTag) {
+        final DetailNode[] childrenOfJavadocInlineTag = javadocInlineTag.getChildren();
+        DetailNode nodeToBeProcessed = childrenOfJavadocInlineTag[2];
```
`current` or `currentNode`? Also, we can move this declaration closer to usage.

Vyom-Yadav replied:
Done

nmancus1 replied:
Resolved.

nmancus1 reviewed (changes requested):
Items:

Vyom-Yadav commented:
@strkkk @nmancus1 Though I feel that second commit shouldn't be minor because-

- Changed Input Files
- Added another Token
- Modified how logic was being processed, decreased time complexity (Didn't calculate complexity through asymptotic notations, used `nanoTime()` method)

Also, the behavior of check has changed.
Config.xml
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="SummaryJavadocCheck"/>
    </module>
</module>
```
Test.java
```
package TestP;

public class Test {

    /**
     * {@summary Missing period}
     */
    void foo() {
    }

    /**
     * {@summary Missing period
     * and having multiple lines}
     */
    void hoo() {
    }
}
```
Earlier behavior-
```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:5: Summary of Javadoc is missing an ending period. [SummaryJavadoc]
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:11: Summary of Javadoc is missing an ending period. [SummaryJavadoc]
Audit done.
Checkstyle ends with 2 errors.
```
Current behaviour-
```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:6: Summary of Javadoc is missing an ending period. [SummaryJavadoc]
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:12: Summary of Javadoc is missing an ending period. [SummaryJavadoc]
Audit done.
Checkstyle ends with 2 errors.
```
The line number has changed but **this is only for errors involving `@summary`**.
I feel this is better as we are shown violation on the line (or in the block) in which it has occurred rather than just pointing at the first line of Javadoc.

Vyom-Yadav commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/summaryjavadoc/InputSummaryJavadocInlineDefault.java:
```diff
@@ -182,8 +173,8 @@ void foo23(){} // ok
-    // violation below
-    /**
+
+    /** // violation below
      * {@summary first sentence is normally the summary.
      * Use of html tags:
      * {@code SomeCodeHere.}
```
`// violation below` can be changed to-
```
// violation 2 lines below
/** 
......
 */
```
after merging #10752

nmancus1 commented:
@Vyom-Yadav is it not possible to make only the simple change outlined in https://github.com/checkstyle/checkstyle/pull/10021#issuecomment-845675251, and maintain current check behavior?

Vyom-Yadav commented:
> @Vyom-Yadav is it not possible to make only the simple change outlined in [#10021 (comment)](https://github.com/checkstyle/checkstyle/pull/10021#issuecomment-845675251), and maintain current check behavior?

@nmancus1 Yes can be done but I don't think it should be done as it will be increasing time-complexity rather than decreasing it. It will be done with-
```
private String getInlineSummary(DetailNode ast) {
        final String inlineSummary;
        final DetailNode inlineNode = JavadocUtil.findFirstToken(ast, JavadocTokenTypes.JAVADOC_INLINE_TAG);
        if(inlineNode == null) {
            DetailNode htmlElementNode = JavadocUtil.findFirstToken(ast, JavadocTokenTypes.HTML_ELEMENT);
            inlineSummary = JavadocUtil.getContentOfInlineCustomTag(getInlineTagNodeWithinHtmlElement(htmlElementNode));
        } else {
            inlineSummary = JavadocUtil.getContentOfInlineCustomTag(inlineNode);
        }
        return inlineSummary;
}
```
But this won't be a good way IMO, this approach searches for the inline tag from the `JavadocTokenTypes.JAVADOC` node and then processes it. This will we redo what https://github.com/checkstyle/checkstyle/blob/eacbf94f8bbad42dbcad3d97d6bf7150b273a20d/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L361 is already doing.

strkkk commented:
> Though I feel that second commit shouldn't be minor because-
> 
>     * Changed Input Files
> 
>     * Added another Token
> 
>     * Modified how logic was being processed, decreased time complexity (Didn't calculate complexity through asymptotic notations, used `nanoTime()` method)
> 
> 
> Also, the behavior of check has changed.

I suggest to move to other PR/issue then. I guess it wont change implementation of the util method, it is only about SummaryJavadocCheck, right?

nmancus1 reviewed (changes requested):
> I will create another issue for it

Please:

1. Create issue, and explain why changes in `SummaryJavadocCheck` are needed.
2. In issue, explain proposed changes to `SummaryJavadocCheck`.
2. Let issue be approved.

I would like to make sure that we are ok to depend on changes being made in `SummaryJavadocCheck` before proceeding here.

Vyom-Yadav commented:
@nmancus1 Created issue at #11030

nmancus1 commented:
@Vyom-Yadav please send PR to contibution repo like https://github.com/checkstyle/contribution/pull/585 to add this check to regression testing configs

Vyom-Yadav commented:
@nmancus1 Just a quick doubt, I added this to `checks-nonjavadoc-error.xml` but as the name of file suggests it checks for non-javadoc errors, a bit confusing, although other javadoc checks were also present in that file.  `SummaryJavadoc` was already present in `checks-only-javadoc-error.xml`.

nmancus1 commented:
@Vyom-Yadav this was supposed to be posted in https://github.com/checkstyle/checkstyle/pull/10930 :)

strkkk commented:
work will continue in https://github.com/checkstyle/checkstyle/issues/11030
